compiler/rustc_mir/src/transform/function_references.rs

```diff
@@ -0,0 +1,158 @@
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::visit::Visitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_session::lint::builtin::FUNCTION_REFERENCES;
+use rustc_span::Span;
+use rustc_target::spec::abi::Abi;
+
+use crate::transform::{MirPass, MirSource};
+
+pub struct FunctionReferences;
+
+impl<'tcx> MirPass<'tcx> for FunctionReferences {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, _src: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let source_info = SourceInfo::outermost(body.span);
+        let mut checker = FunctionRefChecker {
+            tcx,
+            body,
+            potential_lints: Vec::new(),
+            casts: Vec::new(),
+            calls: Vec::new(),
+            source_info,
+        };
+        checker.visit_body(&body);
+    }
+}
+
+struct FunctionRefChecker<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    body: &'a Body<'tcx>,
+    potential_lints: Vec<FunctionRefLint>,
+    casts: Vec<Span>,
+    calls: Vec<Span>,
+    source_info: SourceInfo,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for FunctionRefChecker<'a, 'tcx> {
+    fn visit_basic_block_data(&mut self, block: BasicBlock, data: &BasicBlockData<'tcx>) {
+        self.super_basic_block_data(block, data);
+        for cast_span in self.casts.drain(..) {
+            self.potential_lints.retain(|lint| lint.source_info.span != cast_span);
+        }
+        for call_span in self.calls.drain(..) {
+            self.potential_lints.retain(|lint| lint.source_info.span != call_span);
+        }
+        for lint in self.potential_lints.drain(..) {
+            lint.emit(self.tcx, self.body);
+        }
+    }
+    fn visit_statement(&mut self, statement: &Statement<'tcx>, location: Location) {
+        self.source_info = statement.source_info;
```

Comment:
All these other functions have a `Location` which you can use with [this function](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#method.source_info) to obtain a `SourceInfo`. So I don't think you need to carry that around in the visitor.